fraud/bad_encoding_test.go

```diff
@@ -0,0 +1,161 @@
+package fraud
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	mdutils "github.com/ipfs/go-merkledag/test"
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/pkg/da"
+	"github.com/tendermint/tendermint/pkg/wrapper"
+
+	"github.com/celestiaorg/celestia-node/ipld"
+	"github.com/celestiaorg/celestia-node/service/header"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+func TestCreateBadEncodingFraudProofWithCompletedShares(t *testing.T) {
+	eds := ipld.RandEDS(t, 16)
+	type test struct {
+		name   string
+		isRow  bool
+		roots  [][]byte
+		leaves func(uint) [][]byte
+		length int
+	}
+	tests := []test{
+		{"rowRoots", true, eds.RowRoots(), eds.Row, len(eds.RowRoots())},
+		{"colRoots", false, eds.ColRoots(), eds.Col, len(eds.ColRoots())},
+	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			for index := 0; index < tc.length; index++ {
+				_, err := CreateBadEncodingFraudProof(context.Background(), 1, uint8(index), tc.isRow, eds, tc.roots, tc.leaves(uint(index)), nil)
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestFraudProofValidationForRow(t *testing.T) {
+	dag := mdutils.Mock()
+	eds := ipld.RandEDS(t, 2)
+	size := eds.Width()
+
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(size / 2))
+	_, err := ipld.PutData(context.Background(), ipld.ExtractODSShares(eds), dag)
+	require.NoError(t, err)
+	shares := flatten(eds)
+	copy(shares[3][8:], shares[8][8:])
+	attackerEDS, err := rsmt2d.ImportExtendedDataSquare(shares, rsmt2d.NewRSGF8Codec(), tree.Constructor)
+	require.NoError(t, err)
+	dataSquare := make([][]byte, size*size)
+	copy(dataSquare, shares)
+	dataSquare[2] = nil
+	dataSquare[3] = nil
+	dataSquare[8] = nil
+	dataSquare[12] = nil
+	brokenEDS, err := rsmt2d.RepairExtendedDataSquare(
+		attackerEDS.RowRoots(),
+		attackerEDS.ColRoots(),
+		dataSquare,
+		rsmt2d.NewRSGF8Codec(),
```

Comment:
IMO, we should later test this also with the leopard codec (or we should use the constructor from core everywhere; so the tests will work with whatever library we end up using for launch).